packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -1026,9 +1026,23 @@ class TextSelectionGestureDetectorBuilder {
   // The viewport offset pixels of the [RenderEditable] at the last drag start.
   double _dragStartViewportOffset = 0.0;
 
+  // Returns true iff either shift key is currently down.
+  bool get _isShiftPressed {
+    return HardwareKeyboard.instance.logicalKeysPressed
+      .any(<LogicalKeyboardKey>{
+        LogicalKeyboardKey.shiftLeft,
+        LogicalKeyboardKey.shiftRight,
+      }.contains);
+  }
+
   // True iff a tap + shift has been detected but the tap has not yet come up.
   bool _isShiftTapping = false;
 
+  // For a shift + tap + drag gesture, the TextSelection at the point of the
+  // tap. Mac uses this value to reset to the original selection when an
+  // inversion of the base and offset happens.
+  TextSelection? _shiftTapDragSelection;
```

Comment:
is it true that `_shiftTapDragSelection != null` if and only if `_isShiftTapping` is true? If so then you need only one variable to track the state?

Reply:
It's not true unfortunately.  _isShiftTapping will be true when just shift tapping (not dragging) while _shiftTapDragSelection will be null.